### PR TITLE
Local cache [WIP]

### LINKED ISSUE #173: Local cache for datasets
In order to avoid hammering a network drive when several processes need access to a dataset at the same time, which can be a problem at MILA, it would be really useful to have an automatic way of caching datasets on a local drive.
Pylearn2 has had that functionality for a while [[1](https://github.com/lisa-lab/pylearn2/blob/master/pylearn2/datasets/cache.py)], and it proved useful to reduce the load on file servers.
One way this could work would be:
- Use a new environment variable, for instance `FUEL_DATA_CACHE` to specify the path of that cache. If that variable is not set, no cache will be used.
- In `fuel.utils.find_in_data_path`, when the original data file is found, try to:
  - see if it is already in the cache, and check that the version is up to date,
  - if not, try to copy if into the cache,
  - if it works, return the path of the cached version, otherwise, return the original path.
- Maybe add an option to `find_in_data_path` to disable caching, for instance `ignore_data_cache`. This can be used by a particular Dataset class to disable caching for its data.

CC @nouiz who worked on the Pylearn2 version.

### PR DESCRIPTION
**Pull request by dmitriy-serdyuk**
Continuation of #287

Fixes #173

### CODE REVIEW

File: fuel/utils/cache.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+        # do not locally copy the data
+        if self.dataset_local_dir == "":
+            return filename
+
+        common_msg = ("Message from fuel local cache of dataset"
+                      "(specified by the environment variable "
+                      "FUEL_LOCAL_DATA_PATH): ")
+        # Make sure the file to cache exists and really is a file
+        if not os.path.exists(remote_name):
+            log.error("Error : Specified file %s does not exist" %
```
Comment (rizar) [inline fuel/utils/cache.py]: The `%`-style formatting is not available in Python 3. Please use `str.format`.

File: fuel/utils/common_strings.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+Mac OS X
+========
+Environment variables on Mac OS X work the same as in Linux, except you should
+modify and run the "source" command on ~/.profile rather than ~/.bashrc.
+"""
+
+viewer_command_error_essay = """
+PYLEARN2_VIEWER_COMMAND not defined. PLEASE READ THE FOLLOWING MESSAGE
+CAREFULLY TO SET UP THIS ENVIRONMENT VARIABLE:
+pylearn2 uses an external program to display images. Because different
```
Comment (rizar) [inline fuel/utils/common_strings.py]: Mentions of pylearn2 it the docstring are confusing

File: fuel/datasets/hdf5.py [hunk 3]
```diff
@@ -173,6 +174,7 @@ def __init__(self, file_or_path, which_sets, subset=None,
             self.external_file_handle = file_or_path
         else:
             self.path = file_or_path
+            self.path = dataset_cache.cache_file(self.path)
```
Comment (rizar) [inline fuel/datasets/hdf5.py]: I assume that we should also change the way the files are opened in all other basic dataset classes.

File: fuel/utils/common_strings.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+environment_variable_essay = """
```
Comment (rizar) [inline fuel/utils/common_strings.py]: `essay` is a weird name for a variable.

File: fuel/utils/write_lock.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
+import atexit
```
Comment (rizar) [inline fuel/utils/write_lock.py]: Do we really want to copy all this code to Fuel? Can we not use another library to lock files?

File: fuel/utils/common_strings.py (hunk 2 above)
Comment (rizar) [inline fuel/utils/common_strings.py]: In general, I think that `viewer_command_error_essay` is not used in the cache code.

File: fuel/utils/write_lock.py (hunk 5 above)
Comment (dmitriy-serdyuk) [inline fuel/utils/write_lock.py]: I checked it and I haven't found libraries for that. Stackoverflow mostly suggests to copy-paste a piece of code of questionable quality that doesn't cover all cases. Should we create a small library?

File: fuel/datasets/hdf5.py (hunk 3 above)
Comment (dmitriy-serdyuk) [inline fuel/datasets/hdf5.py]: Hdf5 dataset is the standard one, I decided to start with it.

**Comment by dmitriy-serdyuk** [comment]
Users may share the cache, it depends on the `FUEL_LOCAL_DATA_PATH` variable. The code checks if there is space available. Then the dataset is copied to the local directory and the function returns a new path to the dataset.

CC @nouiz

**Comment by dmitriy-serdyuk** [comment]
Perhaps, we need to add a checksum to the local filename so we can compare if the original file is the same. 

@nouiz , a user may have disc quota smaller than available space, what do we do about that?

**Comment by nouiz** [comment]
On our computers, users don't have quota on the local disk. I didn't see
that. If that happen, I think there is a way to limit the size of the
cache, the user should just use that.

File: fuel/config_parser.py [hunk 6]
```diff
@@ -51,6 +51,11 @@
    The default :class:`~numpy.dtype` to use for floating point numbers. The
    default value is ``float64``. A lower value can save memory.
 
+.. option:: local_data_path
+
+   The local path where the dataset is going to be copied. This is a useful
```
Comment (rizar) [inline fuel/config_parser.py]: Could you please elaborate that the copying will only take place at the first time, and after that the cached copy will be used?

File: fuel/utils/cache.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+    Parameters
+    ----------
+    remote_fname : str
+        Remote file to copy
+    local_fname : str
+        Path and name of the local copy to be made of the remote file.
+
+    """
+
```
Comment (rizar) [inline fuel/utils/cache.py]: Can we add `log.debug` here, so that impatient users could see that their files are being copied?

**Comment by maximecb** [comment]
Following a discussion with Fred, I would like to take over this PR and try to get the feature completed. Will fork dmitriy's fuel repository, fix merge conflicts and open a new PR.

**Comment by rizar** [comment]
Great! I will take a look at Dima's PR to refresh my memory of what was
left to be done.

**Comment by rizar** [comment]
Closed as #400 has been merged.

**Pull request closed without merging**